compiler/rustc_infer/src/infer/canonical/canonicalizer.rs

```diff
@@ -49,6 +49,29 @@ impl<'cx, 'tcx> InferCtxt<'cx, 'tcx> {
         Canonicalizer::canonicalize(value, self, self.tcx, &CanonicalizeAllFreeRegions, query_state)
     }
 
+    /// Like [Self::canonicalize_query], but preserves distinct universes. For
+    /// example, canonicalizing `&'?0: Trait<'?1>`, where `'?0` is in `U1` and
+    /// `'?1` is in `U3` would be canonicalized to have ?0` in `U1` and `'?1`
+    /// in `U2`.
+    pub fn canonicalize_chalk_query<V>(
```

Comment:
This might be better named for what it does `canonicalize_query_preserving_universes` with a comment this is used for Chalk